src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseHeaderMap.cs

```diff
@@ -0,0 +1,256 @@
+            ValidateHeaderValue(label, null, encodedValue);
+
+            _headerParameters[label] = encodedValue;
+        }
+
+        public void SetValue(CoseHeaderLabel label, int value)
+        {
+            ValidateIsReadOnly();
+            ValidateHeaderValue(label, value < 0 ? CborReaderState.NegativeInteger : CborReaderState.UnsignedInteger, null);
+
+            var writer = new CborWriter();
+            writer.WriteInt32(value);
+            _headerParameters[label] = writer.Encode();
+        }
+
+        public void SetValue(CoseHeaderLabel label, string value)
+        {
+            ValidateIsReadOnly();
+            ValidateHeaderValue(label, CborReaderState.TextString, null);
+
+            var writer = new CborWriter();
+            writer.WriteTextString(value);
+            _headerParameters[label] = writer.Encode();
+        }
+
+        public void SetValue(CoseHeaderLabel label, ReadOnlySpan<byte> value)
+        {
+            ValidateIsReadOnly();
+            ValidateHeaderValue(label, CborReaderState.ByteString, null);
+
+            var writer = new CborWriter();
+            writer.WriteByteString(value);
+            _headerParameters[label] = writer.Encode();
+        }
+
+        public void Remove(CoseHeaderLabel label)
+        {
+            ValidateIsReadOnly();
+            _headerParameters.Remove(label);
+        }
+
+        private void ValidateIsReadOnly()
+        {
+            if (_isReadOnly)
+            {
+                throw new InvalidOperationException(SR.CoseHeaderMapDecodedMapIsReadOnlyCannotSetValue);
+            }
+        }
+
+        private void ValidateHeaderValue(CoseHeaderLabel label, CborReaderState? state, ReadOnlyMemory<byte>? encodedValue)
+        {
+            if (state != null)
+            {
+                Debug.Assert(encodedValue == null);
+                if (label.LabelAsInt32 != null)
+                {
+                    Debug.Assert(label.LabelAsString == null);
+                    // all known headers are integers.
+                    ValidateKnownHeaderValue(label.LabelAsInt32.Value, state, null);
+                }
+            }
+            else
+            {
+                Debug.Assert(encodedValue != null);
+                var reader = new CborReader(encodedValue.Value);
+
+                if (label.LabelAsInt32 != null)
+                {
+                    Debug.Assert(label.LabelAsString == null);
+                    // all known headers are integers.
+                    ValidateKnownHeaderValue(label.LabelAsInt32.Value, reader.PeekState(), reader);
+                }
+                else
+                {
+                    Debug.Assert(label.LabelAsString != null);
+                    reader.SkipValue();
+                }
+
+                if (reader.BytesRemaining != 0)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.CoseHeaderMapHeaderDoesNotAcceptSpecifiedValue, label));
```

Comment:
For System.Formats.Asn1:

https://github.com/dotnet/runtime/blob/c7adf5598f27e59644030ddae4e382cbda74e41e/src/libraries/System.Formats.Asn1/src/Resources/Strings.resx#L99-L101

For System.Formats.Cbor:

https://github.com/dotnet/runtime/blob/c7adf5598f27e59644030ddae4e382cbda74e41e/src/libraries/System.Formats.Cbor/src/Resources/Strings.resx#L222-L224

Something along those lines seems more appropriate than Header '{0}' does not accept the specified value.

Reply:
Should we re-think throwing that message in ValidateKnownHeaderValue as well?